curp/src/members.rs

```diff
@@ -1,134 +1,172 @@
+            .map(|t| t.id)
+            .collect()
     }
 
-    /// calculate the member id
-    #[must_use]
-    #[inline]
-    pub fn gen_member_id(&self, cluster_name: &str) -> u64 {
-        let ts = SystemTime::now()
-            .duration_since(SystemTime::UNIX_EPOCH)
-            .unwrap_or_else(|e| unreachable!("SystemTime before UNIX EPOCH! {e}"))
-            .as_secs();
+    /// Calculate the member id
+    fn calculate_member_id(address: &str, cluster_name: &str, timestamp: Option<u64>) -> ServerId {
         let mut hasher = DefaultHasher::new();
-        hasher.write(self.self_address().as_bytes());
+        hasher.write(address.as_bytes());
         hasher.write(cluster_name.as_bytes());
-        hasher.write_u64(ts);
+        if let Some(ts) = timestamp {
+            hasher.write_u64(ts);
+        }
         hasher.finish()
     }
 
-    /// calculate the cluster id
-    #[must_use]
-    #[inline]
-    pub fn gen_cluster_id(&self, cluster_name: &str) -> u64 {
+    /// Calculate the cluster id
+    fn gen_cluster_id(&mut self) {
         let mut hasher = DefaultHasher::new();
-        let cluster_members = self.all_members();
+        for id in self.members.keys().sorted() {
+            hasher.write_u64(*id);
+        }
+        self.cluster_id = hasher.finish();
+    }
 
-        let member_urls = cluster_members
-            .values()
-            .sorted()
-            .map(String::as_str)
-            .collect::<Vec<_>>();
+    /// Get member id
+    #[must_use]
+    #[inline]
+    pub fn self_id(&self) -> ServerId {
+        self.member_id
+    }
 
-        for url in member_urls {
-            hasher.write(url.as_bytes());
-        }
-        hasher.write(cluster_name.as_bytes());
-        hasher.finish()
+    /// Get cluster id
+    #[must_use]
+    #[inline]
+    pub fn cluster_id(&self) -> u64 {
+        self.cluster_id
     }
 
-    /// get peers
+    /// Get peers
     #[must_use]
     #[inline]
     pub fn peers(&self) -> HashMap<ServerId, String> {
-        self.peers.clone()
+        self.members
+            .values()
+            .filter(|t| t.id != self.member_id)
+            .map(|t| (t.id, t.address.clone()))
+            .collect()
     }
 
-    /// get all members
+    /// Get all members
     #[must_use]
     #[inline]
     pub fn all_members(&self) -> HashMap<ServerId, String> {
-        let mut cluster_members = self.peers.clone();
-        let _ignore = cluster_members.insert(self.self_id.clone(), self.self_address.clone());
-        cluster_members
+        self.members
+            .values()
+            .map(|t| (t.id, t.address.clone()))
+            .collect()
     }
 
-    /// peers count
+    /// Get length of peers
     #[must_use]
     #[inline]
     pub fn peers_len(&self) -> usize {
-        self.peers.len()
+        self.members.len().overflow_sub(1)
+    }
+
+    /// Get id by name
+    #[must_use]
+    #[inline]
+    pub fn get_id_by_name(&self, name: &str) -> Option<ServerId> {
```

Comment:
We should be cautious about this method, as there's may multiple members who have the same name. Maybe limit this to test use only.

Reply:
Added `#[cfg(test)]`